sdk/core/src/macros.rs

```diff
@@ -290,6 +290,98 @@ macro_rules! future {
     };
 }
 
+/// The following macro invocation:
+/// ```
+/// create_request_header_cow!(ClientRequestId, CLIENT_REQUEST_ID);
+/// ```
+/// Turns into a Header value used to construct requests.
+#[macro_export]
+macro_rules! create_request_header_cow {
+    ($name:ident, $header:ident $(, $doc:expr)?, $(($variant:ident, $value:expr)), *) => {
+        $( #[doc = $doc] )?
+        #[derive(Debug, Clone)]
+        pub struct $name(std::borrow::Cow<'static, str>);
+
+        impl $name {
+            $(
+                pub const $variant: $name = $name::from_static($value);
+            )*
+
+            pub fn new<S>(s: S) -> Self
+            where
+                S: Into<std::borrow::Cow<'static, str>>,
+            {
+                Self(s.into())
+            }
+
+            pub const fn from_static(s: &'static str) -> Self {
+                Self(std::borrow::Cow::Borrowed(s))
+            }
+        }
+
+        impl<S> From<S> for $name
+        where
+            S: Into<std::borrow::Cow<'static, str>>,
+        {
+            fn from(s: S) -> Self {
+                Self::new(s)
+            }
+        }
+
+        impl $crate::headers::Header for $name {
+            fn name(&self) -> $crate::headers::HeaderName {
+                $crate::headers::$header
+            }
+
+            fn value(&self) -> $crate::headers::HeaderValue {
+                $crate::headers::HeaderValue::from_cow(self.0.clone())
+            }
+        }
+    };
+}
+
+/// The following macro invocation:
+/// ```
+/// create_request_query_cow!(Prefix, "prefix");
+/// ```
+/// Turns into a request query option used to construct requests
+#[macro_export]
+macro_rules! create_request_query_cow {
+    ($name:ident, $option:expr $(, $doc:expr)?) => {
```

Comment:
Can we use `($(#[$doc:meta])*` for the docs so that we don't have to write docs as string literals?